src/integrations/feed/clients.ts

```diff
@@ -0,0 +1,167 @@
+      retries: 5,
+    });
+    if (!res.data.length) {
+      ctx?.log.warn({ config }, 'empty response received from feed service');
+      return [];
+    }
+    return res.data.map(({ post_id, metadata }) => [
+      post_id,
+      (metadata && JSON.stringify(metadata)) || null,
+    ]);
+  }
+}
+
+/**
+ * Feed client that caches the response in Redis
+ */
+export class CachedFeedClient implements IFeedClient {
+  private readonly client: IFeedClient;
+  private readonly redisPool: IORedisPool;
+
+  constructor(client: IFeedClient, redisPool: IORedisPool) {
+    this.client = client;
+    this.redisPool = redisPool;
+  }
+
+  getCacheKeyPrefix(feedId?: string): string {
+    return `feeds:${feedId || 'global'}`;
+  }
+
+  getCacheKey(userId?: string, feedId?: string): string {
+    return `${this.getCacheKeyPrefix(feedId)}:${userId || 'anonymous'}`;
+  }
+
+  async shouldServeFromCache(
+    offset: number,
+    key: string,
+    feedId?: string,
+  ): Promise<boolean> {
+    if (offset) {
+      return true;
+    }
+    const updateKey = `${this.getCacheKeyPrefix(feedId)}:update`;
+    const [lastGenerated, lastUpdated] = await this.redisPool.execute(
+      async (client) => {
+        return client.mget(`${key}:time`, updateKey);
+      },
+    );
+    const ttl = 3 * 60 * 1000;
+    return !(
+      !lastGenerated ||
+      (lastUpdated && lastUpdated > lastGenerated) ||
+      new Date().getTime() - new Date(lastGenerated).getTime() > ttl
+    );
+  }
+
+  async fetchFromCache(
+    ctx: Context,
+    feedId: string | undefined,
+    config: FeedConfig,
+  ): Promise<FeedResponse | undefined> {
+    const { offset, page_size: pageSize } = config;
+    const key = this.getCacheKey(config.user_id, feedId);
+    const cachePromise = this.redisPool.execute(async (client) => {
+      return client.get(`${key}:posts`);
+    });
+    if (await this.shouldServeFromCache(offset, key, feedId)) {
+      const cachedFeed = JSON.parse(await cachePromise);
+      if (cachedFeed?.length) {
+        const page: string[] | [string, string | undefined][] =
+          cachedFeed.slice(offset, pageSize + offset);
+        // Support legacy cache that contains only post id
+        if (page.length && typeof page[0] === 'string') {
+          return (page as string[]).map((postId) => [postId, undefined]);
+        }
+        return page as [string, string | undefined][];
+      }
+    }
+    return undefined;
+  }
+
+  async updateCache(
+    feedId: string,
+    config: FeedConfig,
+    items: FeedResponse,
+  ): Promise<void> {
+    const key = this.getCacheKey(config.user_id, feedId);
+    await this.redisPool.execute(async (client) => {
+      const pipeline = client.pipeline();
+      pipeline.set(
+        `${key}:time`,
+        new Date().toISOString(),
+        'EX',
+        ONE_DAY_SECONDS,
+      );
+      pipeline.set(
+        `${key}:posts`,
+        JSON.stringify(items),
+        'EX',
+        ONE_DAY_SECONDS,
```

Comment:
I seen this before all around, maybe not applicable for this but while we are refactoring maybe create a key generator to abstract this syntax to something easier to repeat all around

Reply:
Cache is gonna move to feed service side, so I don't want to introduce any changes here. I copied it as it was before. Wdyt?

Reply:
Yes, that works of course